### PR TITLE
WIP - [DF] Implement a C++ workflow generator for DistRDF workers

### PR DESCRIPTION
**Pull request by etejedor**
This is a first stab to an optimized DF workflow generator in the
DistRDF workers. The optimization consists in generating the
C++ code of the workflow, compile it with ACLiC and execute it.
A compiled workflow might be faster than its purely jitted version.

The optimized mode coexists with the old (jitted) mode, which is
the default. To enable the optimized mode the following flag
can be used:
```python
ROOT.RDF.Experimental.Distributed.optimized = True
```

### CODE REVIEW

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (vepadulano) [inline bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:345]: This is very nice and true for multiprocessing, not sure about multithreading. I'd say we need at least:
* a lock on the file being written
* the first thread arriving to this function needs to compile the function before other threads so it can fill the cache of the class 

Although by looking at the requirements above it seems to me this mechanism would need to happen just once per computation graph per worker ? First process/thread that starts this compilation locks a file with the name being just the hash of the code so all other processes/threads would be waiting for that file to unlock? Not sure

**Review by vepadulano — COMMENTED** [review]
Thanks a lot Enric! Can't wait to try this out :smile:

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:139]: Nice thanks!

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:296]: `narg` here is the argument index/number, I add the comma before the argument only if for `narg>0` (so starting from the second argument, if any). If there are no arguments at all of course there is no iteration.

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:363]: A hash can be a negative integer value (thus with a `-` in its string representation). I replace `-` by `_` so that the hash can be used as part of the workflow generator function name (`-` is not allowed in function names).

It's not obvious, I added a comment to explain this :)

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:345]: As you can see I have a ` TODO: Make this function thread-safe? To support Dask threaded workers` at the beginning of this function :smile: 

Right now the cache is just reused by the same worker process that wants to run the same workflow on another fragment of the dataset, in the same application run. It could also be the case, if the worker process survives across different application runs, that it reuses a cache from a previous run (because the user submitted a second job with the exact same workflow).

If we want to support thread-safe caching, we need to have a look at the alternatives for locking a file from Python. It seems there are a few packages that provide advisory locking features (which is good enough for us), you can have a read at [this thread](https://stackoverflow.com/questions/489861/locking-a-file-in-python) for a discussion on this topic.

However, I would leave this for a second round, since testing the caching is not what has highest priority now: it is testing the performance of the optimized workflow generation vs the old way (with new jitting optimizations).

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (vepadulano) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:365]: Alright nice thanks!

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (vepadulano) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:296]: Ah yes now I see thanks :+1:

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (vepadulano) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:345]: Yes I saw your comment, I just wanted to start a conversation on this because I think it is a very interesting part of this feature :smile: I agree it can be left for a second iteration :+1:

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (eguiraud) [inline bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:263]: if I understand correctly, we increment `res_ptr_id` here because we know that _eventually_ we will insert a result at position `i` for this action. Until then, indexing into the list of results using `res_ptr_id` might be tricky, e.g. we have to make sure to add the result for this action in the list before we do a substitution of the results of Snapshot.

This implicit required sequencing seems easy to forget and break when refactoring, could we instead already add dummy contents to the relevant vectors so we later we only have to substitute contents and it doesn't matter in which order we do it?

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (eguiraud) [inline bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:375]: if I understand correctly this is relying on the fact that the first occurrence of `GenerateGraph` in `code` (also as a part of longer identifiers) is at the point of the declaration. It's probably a safe bet, but maybe even safer would be to put something like `__RDF_FUNC_NAME_PLACEHOLDER__` in `code` and replace that?

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (eguiraud) [inline bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:366]: this fails badly (and possibly silently) in the (very) rare case of a hash collision. hashtables, sets, dicts etc. typically have code to deal with that case, do we need some too?

**Review by eguiraud — CHANGES_REQUESTED** [review]
This is amazing and it was a nice read :D thank you for the many useful code comments.

The only things really missing are tests, a nullptr check, and maybe an improvement to the Snapshot-laziness (so that we don't ignore user-provided Snapshot options). The rest of the comments are just (mostly minor) suggestions, several of which you can simply ignore if you don't agree with them.

**Comment by eguiraud** [comment]
@Axel-Naumann @etejedor tangentially related, could we require Python 3 for experimental stuff, similarly to how we require C++17 for ROOT7? It would be great to be able to use f-strings in this PR, for example.

**Comment by etejedor** [comment]
@eguiraud I'd say yes. I think it's a good compromise: PyROOT in general and local RDF would still support Python2, but the distributed features are written for Python3. This piece of code, though, might end up being used also in local RDF, so I'd keep it as it is for now.

File: bindings/experimental/distrdf/python/DistRDF/ComputationGraphGenerator.py
Comment (vepadulano) [reply bindings/experimental/distrdf/python/DistRDF/ComputationGraphGenerator.py:18]: The `logging` format is left to the user in Python. At the time I first implemented this I followed the guide in the [Python docs](https://docs.python.org/3/howto/logging.html#advanced-logging-tutorial) that states some good practices. Modules that want to log something should use `logger = logging.getLogger(__name__)` so that whenever the user will write their logger in their application they will at least get something like 
```DistRDF.ComputationGraphGenerator: ...```
Plus any other formatting option. For simplicity, DistRDF offers a predefined logger through `DistRDF.create_logger` so that the user will get something like
```
DEBUG: DistRDF.Proxy[2021-09-08 12:19:42,275]: Created new Count node
```
in their output. It is still worth tracking this discussion in a separate issue, since I believe at some point we want to use ROOT logging system for DistRDF logs right?

Comment (eguiraud) [reply bindings/experimental/distrdf/python/DistRDF/ComputationGraphGenerator.py:18]: this a general PyROOT question, but imho we should not mix different logging mechanisms until a decision is taken

Comment (vepadulano) [reply bindings/experimental/distrdf/python/DistRDF/ComputationGraphGenerator.py:18]: I totally agree! This logging mechanism is a remnant from PyRDF

File: bindings/experimental/distrdf/python/DistRDF/Backends/Base.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/Backends/Base.py:246]: I think we discussed this in person and we said either option was fine? Option1: implementing `GetResultPtr` in `RResultHandle`, Option2: what you just suggested.

File: bindings/experimental/distrdf/python/DistRDF/ComputationGraphGenerator.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/ComputationGraphGenerator.py:18]: I agree with having a uniform logging system. I'll leave this logging message as it is now, more as a placeholder than anything else. When we change it for the whole DistRDF we can change this instance too.

File: bindings/experimental/distrdf/python/DistRDF/ComputationGraphGenerator.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/ComputationGraphGenerator.py:193]: Perhaps yes, since the calls to `add_node` receive a `parent_idx`, which implies some knowledge about the existence of a graph. Otherwise it would just be fine to traverse the graph outside and add new operations individually.

So I'll move this graph traversal to `__init__` of `CppWorkflow` - still I'll flag the individual `add_node`, `add_lambda`, etc. operations as public, just in case we need to do individual additions later (e.g. from local RDF).

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:175]: I take the first suggestion for the order change. Regarding the second suggestion, I think it's fine to have an in/out parameter (it's in the docs of `_handle_snapshot`, and with your comment the reader can see it already here).

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:205]: In what case ROOT wouldn't be able to return a TClass for an `RResultPtr<T>`?

Comment (eguiraud) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:205]: in case the interpreter does not know `T`

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:212]: Ok! But `Snapshot` handling needs to happen before the call is constructed, and `AsNumpy` can just be at the beginning and do the early return, why move them down?

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:264]: Yes your reasoning is right, the insertion of the AsNumpy results needs to happen before the Snapshots are treated. And yes indeed I thought of the problem you just described, but I'd need to create an empty `RResultHandle` constructor just for that (the AsNumpy dummies). If you think that's preferable, I can change it.

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:315]: Yes the old logic took care of this, still I would just leave this as it is for now. Handling this properly here requires getting the C++ pointer from the Python proxy and use that address in the C++ workflow (I think I can do this with `AddressOf` but I need to try). This is true not only for Snapshot options, but for any other object that RDF calls receive.

Anyway, merging this feature does not change anything unless you activate the `optimized` option. I'm more interested in doing a performance comparison of a first version of this to see if it's even worth investing more time in perfecting it.

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:367]: It would be really really rare, also taking into account that these caches will be rather short-lived. I'll add this as a TODO but not for this PR (for the same reasons I just described above).

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:376]: Yes ok!

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:418]: Yes, if the copy is cheap I can just construct copies when creating `results` with a list comprehension.

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:488]: What's the advantage of one over the other here?

File: bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py
Comment (etejedor) [reply bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py:64]: I see your point, but this would cause a repetition of this code that I don't like. Note that when the action is instant (default), the import error would still happen at `AsNumpy` call time. But if you declare it lazy, everything happens lazily (even the error). I don't think that's too unfair.

File: bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py
Comment (etejedor) [reply bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py:109]: I asked myself what to do with attributes in this class and also in `CppWorkflow` w.r.t. naming conventions. I find it a bit annoying to have to prepend every data attribute with `_` if you only want to access them via methods. In `CppWorkflow` I decided to only flag as public/private the methods and leave alone the data attributes. I can add the `_` flag here since it's a class that will be exposed to the user - to be discussed what we do with RDF's internal classes.

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:205]: Yes, but there is a restricted set of `T` that can go in `RResultPtr<T>` no? Depending on the particular action. Shouldn't ROOT know all those?

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (eguiraud) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:315]: i totally agree on making the performance comparison and verifying that this is worth pursuing before fixing this -- but both things can happen before merging the feature :D 

(or we merge a feature with a known bug, but what's the benefit?)

File: bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py
Comment (eguiraud) [reply bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py:109]: >  I find it a bit annoying to have to prepend every data attribute with `_`

if they are internal, that's what PEP 8 prescribes.

again alternatively we can just make it so that it's ok to access the attributes so we can keep them public.

File: bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py
Comment (eguiraud) [reply bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py:64]: > this would cause a repetition of this code that I don't like

i was not suggesting to repeat the check but to only have it here, where `AsNumpy` is called, is that not enough?

File: bindings/experimental/distrdf/python/DistRDF/ComputationGraphGenerator.py
Comment (eguiraud) [reply bindings/experimental/distrdf/python/DistRDF/ComputationGraphGenerator.py:18]: is there an issue to track this?

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (eguiraud) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:205]: some actions can return arbitrary `T`s. I don't know if any of those are supported by distRDF at the moment, but they might be in the future. but also and most importantly I learnt that I cannot foresee all edge cases in the ROOT I/O system :smile: better play it safe

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (eguiraud) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:418]: yes it's just a few pointers, copy is cheap

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (eguiraud) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:212]: you're right, let's leave them up :smile:

File: bindings/experimental/distrdf/python/DistRDF/Backends/Base.py
Comment (eguiraud) [reply bindings/experimental/distrdf/python/DistRDF/Backends/Base.py:246]: ok yes, this is actually less overall work

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (eguiraud) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:264]: you don't _need_ an empty RResultHandle constructor, you can construct a RResultHandle from an empty RResultPtr instead -- anyway why not: https://github.com/root-project/root/pull/8979

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (eguiraud) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:488]: no practical advantage here.

`using` is the superior alternative so that's what's used by default in modern C++ ("superior" means mostly just that it supports template type aliases).

(interestingly, as a consequence `typedef` often sends the message to the reader that they are reading old code)

File: bindings/experimental/distrdf/python/DistRDF/ComputationGraphGenerator.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/ComputationGraphGenerator.py:18]: Yes https://github.com/root-project/root/issues/8982

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:205]: Ok, I'll throw an exception from the C++ kernel code then.

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:264]: I think it's better this way, otherwise we'd be creating a dummy inside a dummy.

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:315]: Since it's experimental, I don't see any problem in merging something that we know does not work for all cases (and it's not even enabled by default).

Just for the sake of (finer) granularity of the PR review process, I would like to have this in, test it, see where we are in terms of performance, then implement the compiled lambdas on top, another PR, review it, merge it, repeat performance tests. And so on and so forth for any later optimization we might think of.

And if at any stage we decide it's not worth pursuing anymore, we just kill it, it's experimental.

File: bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py
Comment (etejedor) [reply bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py:109]: No I think it's better that they just stay private - I'll flag them as such.

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (eguiraud) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:315]: we don't need the feature in upstream master in order to do testing and performance benchmarking.
the experimental namespace means "hey users, be warned that this API might change in the future", not "let's see what happens if we merge this, we can revert later if we change our mind".

with that said if you prefer merging I would suggest to at least print a warning if Snapshot is called with an options parameter and `distributed.optimized == True`.

File: bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py
Comment (etejedor) [reply bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py:64]: I still think that laziness also means complaining about missing stuff when it's actually required, but ok, I'll leave the try except with numpy here and the two imports unprotected down in GetValue.

Comment (eguiraud) [reply bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py:64]: > I still think that laziness also means complaining about missing stuff when it's actually required

Yes, absolutely, but we _know_ numpy will be required the moment users call `AsNumpy`, right?

Comment (etejedor) [reply bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py:64]: If we are strict that's not completely true, because it could happen that the result of `AsNumpy` is never claimed. In that case, numpy is actually never needed.

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:367]: As discussed: to be reimplemented with a dictionary where the key is the C++ kernel code and the value is some ID that is included in the cache file name.

File: bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py
Comment (etejedor) [reply bindings/experimental/distrdf/python/DistRDF/CppWorkflow.py:315]: I do think that experimental also means we can change our mind. And like I said, this won't even be visible to users since its activation depends on a (now opaque) option.

Anyway, if we did what you suggest, we'd better move these PRs to a branch in my repo, instead of doing them to ROOT master. That would also allow the step-by-step review+merge+test cycles that I described.

I don't think that's necessary though, I'll add the warning.

File: bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py
Comment (eguiraud) [reply bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py:64]: ah come on now :smile: let's code for the common usecase

Comment (etejedor) [reply bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py:64]: If I construct an RDF with a wrong tree name, that's fine until I do something with that RDF, then I see the error. Is that conceptually different than the case here?

Comment (eguiraud) [reply bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py:64]: I don't know where you are going here, don't you agree that erroring out early and at the point of call is more helpful and simpler for users than erroring out from a different place and possibly much later than when the problem was introduced?

(there are considerations that make it trickier for RDF's constructor than for the case here, but sure, in general it would be nice to tell users "hey, you are saying that you want to process this tree and file, but they don't exist" at the point they construct the RDF)

Comment (etejedor) [reply bindings/pyroot/pythonizations/python/ROOT/pythonization/_rdataframe.py:64]: I'll move it, it's fine. I don't have the same perception as you that erroring out at the moment when NumPy is actually needed is bad, since we are working with futures and laziness. But I agree that the early message can help users.

**Comment by etejedor** [comment]
All comments by @eguiraud have been addressed, @vepadulano please approve!

File: tree/dataframe/src/RDFHelpers.cxx
Comment (vepadulano) [inline tree/dataframe/src/RDFHelpers.cxx:50]: I'm not sure this should go in this commit/PR, but it's surely better to check than not

**Review by vepadulano — APPROVED** [review]

File: tree/dataframe/src/RDFHelpers.cxx
Comment (etejedor) [reply tree/dataframe/src/RDFHelpers.cxx:50]: Indeed, it should have been modified together with the inclusion of the default constructor for `RResultHandle`, but it's justified that it goes together with this commit: since I'm creating empty `RResultHandle`s for `AsNumpy` actions, this check is necessary to run `RunGraphs` (otherwise it segfaults).

**Pull request merged**
